### PR TITLE
Fb1228

### PR DESCRIPTION
**Pull request by spireken**

### CODE REVIEW

File: Include/Ashkal/Ashkal/IntersectShapeVisitor.hpp [hunk 1]
```diff
@@ -0,0 +1,18 @@
+#ifndef ASHKAL_INTERSECT_SHAPE_VISITOR_HPP
+#define ASHKAL_INTERSECT_SHAPE_VISITOR_HPP
+#include "Ashkal/Ashkal/Point.hpp"
+#include "Ashkal/Shapes/Shapes.hpp"
+
+namespace Ashkal {
+  //! Determines if a shape is intersecting a Rectangle.
```
Comment (kranar) [inline Include/Ashkal/Ashkal/IntersectShapeVisitor.hpp:8]: Place blank line above documentation, ie. above line 7.

File: Include/Ashkal/Shapes/Ellipse.hpp [hunk 2]
```diff
@@ -7,25 +7,25 @@ namespace Ashkal {
...
+      //! Constructs an Ellipse with radius 0.
       Ellipse();
 
-      //! Constructs an Ellipse with a defined width and height.
+      //! Constructs an Ellipse with a defined major and minor radius.
       /*
-        \param width The width of the Ellipse.
-        \param height The height of the Ellipse.
+        \param major The major radius of the Ellipse.
+        \param mijor The minor radius of the Ellipse.
```
Comment (kranar) [inline Include/Ashkal/Shapes/Ellipse.hpp]: Fix typo ```minor```.

File: Source/Common/IntersectShapeVisitor.cpp [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+    return y_intercept / slope;
+  }
+
+  double slope(const Line& line, const Point& line_pos) {
+    auto point_pos = Point{line.get_point().x + line_pos.x,
+      line.get_point().y + line_pos.y};
+    return (line_pos.y - point_pos.y) / (line_pos.x - point_pos.x);
+  }
+
+  bool within_line(double line_point, double line_pos, double value) {
```
Comment (kranar) [inline Source/Common/IntersectShapeVisitor.cpp]: Boolean functions start with `is_*` or `has_*`.

File: Source/Common/IntersectShapeVisitor.cpp [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+    auto high = std::max(line_point, line_pos);
+    auto low = std::min(line_point, line_pos);
+    if(std::clamp(value, low, high) == value) {
```
Comment (kranar) [inline Source/Common/IntersectShapeVisitor.cpp]: Simply return the expression: ie

```return std::clamp(...) == ...;```

File: Source/Common/IntersectShapeVisitor.cpp [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+    }
+    return false;
+  }
+
+  std::array<std::pair<Line, Point>, 4> get_lines_from_rect(
+      const Rectangle& rect, const Point& pos) {
+    auto lines = std::array<std::pair<Line, Point>, 4>();
+    lines[0] = std::pair<Line, Point>(Line({rect.get_width(), 0}), pos);
+    lines[1] = std::pair<Line, Point>(Line({0, -rect.get_height()}),
+          Point{pos.x + rect.get_width(), pos.y});
```
Comment (kranar) [inline Source/Common/IntersectShapeVisitor.cpp]: Only single indent.

Comment (kranar) [reply Source/Common/IntersectShapeVisitor.cpp]: Use template deduction guides to remove the ```<Line, Point>```.

ie.

```lines[x] = std::pair(a, b);```;

File: Source/Common/IntersectShapeVisitor.cpp [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+    lines[2] = std::pair<Line, Point>(Line({rect.get_width(), 0}),
+          Point{pos.x, pos.y - rect.get_height()});
+    lines[3] = std::pair<Line, Point>(Line({0, -rect.get_height()}), pos);
+    return lines;
+  }
+
+  bool ellipse_intersects(const Line& line, const Point& line_pos,
+    const Ellipse& ellipse, const Point& ellipse_pos) {
```
Comment (kranar) [inline Source/Common/IntersectShapeVisitor.cpp]: Double indent line 50.

File: Source/Common/IntersectShapeVisitor.cpp [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+    auto s = slope(line, line_pos);
+    auto major = ellipse.get_major_radius();
+    auto minor = ellipse.get_minor_radius();
+    auto lcm = std::lcm(static_cast<int>(major), static_cast<int>(minor));
```
Comment (kranar) [inline Source/Common/IntersectShapeVisitor.cpp]: This is likely a bug, casting a ```double``` to ```int``` to get the ```lcm``` seems suspicious.

File: Source/Common/IntersectShapeVisitor.cpp [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+    auto a = (lcm / major) + (lcm / minor);
+    auto b = ((-ellipse_pos.x * 2) * (lcm / major)) +
+      ((-ellipse_pos.y * 2) * (lcm / minor));
+    auto c = ((ellipse_pos.x * ellipse_pos.x * (lcm / major)) +
+      (ellipse_pos.y * ellipse_pos.y) * (lcm / minor)) - std::pow(lcm, 2);
+    auto value = (-b + std::sqrt(std::pow(b, 2) - (4 * a * c))) / (2 * a);
```
Comment (kranar) [inline Source/Common/IntersectShapeVisitor.cpp]: Rename ```value``` to ```root```.

File: Source/Common/IntersectShapeVisitor.cpp [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+  bool rect_contains(const Rectangle& rect, const Point& rect_pos,
+      const Point& point) {
+    return rect_contains_x(rect, rect_pos, point.x) &&
+      rect_contains_y(rect, rect_pos, point.y);
+  }
+
+  bool line_intersects(const Line& line1, const Point& line1_pos,
+      const Line& line2, const Point& line2_pos,
+      const Rectangle& rect, const Point& ref_pos) {
+    double slope1;
```
Comment (kranar) [inline Source/Common/IntersectShapeVisitor.cpp]: Declare these at the point of initialization.

File: Source/Common/IntersectShapeVisitor.cpp [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+    if(std::isinf(slope1) && std::isinf(slope2)) {
+      if(line1_pos.x == line2_pos.x && rect_contains_x(rect, ref_pos,
+          line1_pos.x)) {
+        return true;
+      }
+      return false;
+    }
+    auto y_intercept1 = y_intercept(line1_pos, slope1);
+    auto y_intercept2 = y_intercept(line2_pos, slope2);
+    double x;
```
Comment (kranar) [inline Source/Common/IntersectShapeVisitor.cpp]: Declare this at the point of initialization.

File: Source/Common/IntersectShapeVisitor.cpp [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+      m_intersects = false;
+      m_rect = a;
+      m_rect_pos = p1;
+      m_shape_pos = p2;
+      b.accept(*this);
+      return m_intersects;
+    }
+
+    void visit(const Ellipse& ellipse) override {
+      for(auto line : get_lines_from_rect(m_rect, m_rect_pos)) {
```
Comment (kranar) [inline Source/Common/IntersectShapeVisitor.cpp]: ```auto&```

File: Source/Common/IntersectShapeVisitor.cpp [hunk 12]
```diff
@@ -0,0 +1,260 @@
...
+          rect_contains(m_rect, m_rect_pos, bottom_right)) {
+        m_intersects = true;
+        return;
+      }
+      auto point = get_lines_from_rect(m_rect, m_rect_pos)[0].second;
+      auto n1 = std::pow(point.x - m_shape_pos.x, 2);
+      auto n2 = std::pow(point.y - m_shape_pos.y, 2);
+      auto d1 = std::pow(ellipse.get_major_radius(), 2);
+      auto d2 = std::pow(ellipse.get_minor_radius(), 2);
+      auto lcm = std::lcm(static_cast<int>(d1), static_cast<int>(d2));
```
Comment (kranar) [inline Source/Common/IntersectShapeVisitor.cpp]: Once again, highly suspicious.

File: Source/Common/IntersectShapeVisitor.cpp [hunk 13]
```diff
@@ -0,0 +1,260 @@
...
+    }
+
+    void visit(const Line& line) override {
+      auto pos_x = line.get_point().x + m_shape_pos.x;
+      auto pos_y = line.get_point().y + m_shape_pos.y;
+      if(rect_contains(m_rect, m_rect_pos, m_shape_pos) ||
+          rect_contains(m_rect, m_rect_pos, Point{pos_x, pos_y})) {
+        m_intersects = true;
+      }
+      for(auto rect_line : get_lines_from_rect(m_rect, m_rect_pos)) {
```
Comment (kranar) [inline Source/Common/IntersectShapeVisitor.cpp]: ```auto&```

File: Source/Common/IntersectShapeVisitor.cpp [hunk 14]
```diff
@@ -0,0 +1,260 @@
...
+      m_intersects = true;
+    }
+
+    void visit(const Triangle& triangle) override {
+      auto& points = triangle.get_points();
+      auto lines = std::array<std::pair<Line, Point>, 3>();
+      lines[0] = {Line({points[0]}), Point{points[1]}};
+      lines[1] = {Line({points[1]}), Point{points[2]}};
+      lines[2] = {Line({points[0]}), Point{points[2]}};
+      for(auto& l : lines) {
```
Comment (kranar) [inline Source/Common/IntersectShapeVisitor.cpp]: Destructure this as follows:

```for(auto& [line, point] : lines) { ... }```

File: Source/Common/IntersectShapeVisitor.cpp [hunk 15]
```diff
@@ -0,0 +1,260 @@
...
+        auto& line = l.first;
+        auto& point = l.second;
+        if(slope(line, point) > 0) {
+          if(line.get_point().x > point.x) {
+            line = Line({line.get_point().x - point.x,
```
Comment (kranar) [inline Source/Common/IntersectShapeVisitor.cpp]: What's the difference between these two branches? They look the same to me.

File: Source/Common/IntersectShapeVisitor.cpp [hunk 16]
```diff
@@ -0,0 +1,260 @@
...
+              point.y - line.get_point().y});
+          } else {
+            line = Line({line.get_point().x - point.x,
+              line.get_point().y - point.y});
+          }
+        } else {
+          if(line.get_point().x > point.x) {
+            line = Line({line.get_point().x - point.x, 0});
```
Comment (kranar) [inline Source/Common/IntersectShapeVisitor.cpp]: These two branches look identical.

File: Source/Common/IntersectShapeVisitor.cpp [hunk 17]
```diff
@@ -0,0 +1,260 @@
...
+          } else {
+            line = Line({line.get_point().x - point.x, 0});
+          }
+        }
+      }
+      for(auto& tri_line : lines) {
+        for(auto rect_line : get_lines_from_rect(m_rect, m_rect_pos)) {
```
Comment (kranar) [inline Source/Common/IntersectShapeVisitor.cpp]: ```auto&```

**Review by kranar — CHANGES_REQUESTED** [review]
Very nice man, huge progress.

**Comment by spireken** [comment]
This is mostly just cleanup and style changes, I still have to remove the uses of std::lcm().

**Pull request merged**